numpy/lib/mixins.py

```diff
@@ -0,0 +1,167 @@
+"""Mixin classes for custom array types that don't inherit from ndarray."""
+from __future__ import division, absolute_import, print_function
+
+import sys
+
+from numpy.core import umath as um
+
+# None of this module should be exposed in top-level NumPy module.
+__all__ = []
+
+
+def _binary_method(ufunc):
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', self, other)
```

Comment:
I do like the idea of `__array_ufunc__ = None` disallowing ufunc overrides from any other arguments.

One additional argument in favor of directly calling the ufunc is that, per http://bugs.python.org/issue30140, Python's binary op dispatch is broken for subclasses. So calling the ufunc directly will allow simpler, less buggy `__array_ufunc__` implementations.